patroni/postgresql/__init__.py

```diff
@@ -984,16 +984,17 @@ def pick_synchronous_standby(self, cluster, sync_node_count=1):
         # receiving changes faster than the sync member (very rare but possible). Such cases would
         # trigger sync standby member swapping frequently and the sort on sync_state desc should
         # help in keeping the query result consistent.
-        for app_name, state, sync_state in self.query(
+        for app_name, state, sync_state, replica_rank in self.query(
                 "SELECT pg_catalog.lower(application_name), state, sync_state"
+                ", rank() over (order by {0}_{1} DESC) as replica_rank"
```

Comment:
It seems that we are back to roots, i.e. `ORDER BY flush_lsn DESC, sync_state DESC`.

The issue with the `RANK()` and `replica_rank == 1` check that it would require all replicas to be on the same LSN, which IMO on the system with a lot of TPS would rarely be the case.

Reply:
Agree. We could probably do the opposite, maybe identifying unhealthy sync (if the replication delay is over the threshold) and taking it out from `sync_nodes` list?

Reply:
There is no guaranty that `replay/flush/write` for sync replicas would have the same values.
It is very likely that soon or later Patroni will catch the moment when different replicas are off from each other by a few bytes/kilobytes, therefore not all of them will have the same rank.

> maybe identifying unhealthy sync (if the delay is over the threshold) and taking it out from sync_nodes list?

It could be an option, there are a few questions though:
1. What would be the threshold? Should it be configurable?
2. If there are many replicas the threshold could be applied from the most up-to-date. What should we do if there is only one replica? Should we use the current wal lsn of the primary?

Reply:
Thanks. The safest approach would be going with identifying unhealthy synchronous replica and demoting it to asynchronous:

1. Yes, This will be a configurable parameter, with an acceptable default value.
2. Do you foresee any issues using current lsn of the primary in case of both, i.e. multiple replicas and single replica?

Reply:
Transactions are first always written to the WAL on the primary and only after that they are sent to replicas.
New transactions don't necessarily wait until other transactions are confirmed by sync standbys.
If the transaction rate on the primary is high enough, the current lsn would be most likely is ahead of feedback values reported via pg_stat_replication. If the threshold is too low that would become a problem, because Patroni will switch to async replication.

Reply:
Agree & Your suggestion above makes sense in this case. For a single replica, we could use current wal lsn of the primary. Will start making changes. Thanks.